### PR TITLE
[dali] Initial contribution

### PR DESCRIPTION
**Pull request by rs22**
This PR includes initial support for controlling DALI lighting devices over a Tridonic/Lunatone [DALI USB](https://www.lunatone.com/en/product/dali-usb/) interface with an intermediary [daliserver](https://github.com/onitake/daliserver) that turns the USB HID interface of the adapter into a TCP server.

The binding can be used to control the dimming level of individual devices as well as device groups.
Furthermore, I've added an 'rgb' thing that consists of three individual DALI addresses representing r/g/b LED channels in order to be able to set the composite lighting color in OpenHAB.

This is my first time contributing to this project, so please let me know if somewhere I didn't follow the conventions and coding standards of the project.

There is a forum thread about DALI: https://community.openhab.org/t/lighting-dali-led-strips-and-many-questions/46039/4

### CODE REVIEW

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java:33]: Please remove this lib (and all apache.commons) as currently it is getting removed from all bindings and project, see the sticky on the issues page.

File: bundles/org.openhab.binding.dali/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/resources/OH-INF/thing/thing-types.xml:60]:
```suggestion
		<category>Light</category>
```
DimmableLight does not exist despite it getting used by some bindings in the past. I believe they were removed a month ago. In OH3 you can see the ones when you EDIT an item, there is a searchable list.

File: bundles/org.openhab.binding.dali/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/resources/OH-INF/thing/thing-types.xml:45]: All channels should use mixedCase naming convention.

dimImmediately

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/protocol/DaliFrame.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/protocol/DaliFrame.java:19]: remove apache.commons

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java:153]: is something missing from here? if no please remove the todo.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java:106]: config.port == 0
This should be impossible as you have defined the port in the XML that it must be 1 as a minimum. The framework will enforce this.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java:76]: Is code missing from here?

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java:93]: See https://www.openhab.org/docs/developer/guidelines.html#f-logging
Is HEARTBEAT_MINUTES a constant? Does this need to be logged? If so then it probably should be DEBUG or TRACE

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java:77]: See https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java:152]: https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java
Comment (rs22) [reply bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java:77]: From the docs: `// DONT, DONT, really DONT do that` -- yeah, now the message also got through to me... :pensive: 
Seriously, I seem to have picked the [wrong binding](https://github.com/openhab/openhab-addons/blob/48a5a17b8188ad18d1094a579156cfaccf4dbecc/bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyShadeHandler.java#L68) as the blueprint for this code.

**Comment by rs22** [comment]
@Skinah Thanks for your review!

File: bundles/org.openhab.binding.dali/README.md
Comment (Skinah) [inline bundles/org.openhab.binding.dali/README.md:34]: I find the double up of information a little confusing at first. Parameter and Paramter ID and I'm guessing it is because the main UI gives a nice label VS textual config gives a camelCase version. I think having both could make someone try and use 'Hostname' instead of `host` in a texual file and then wonder why it does not work.
Usually bindings just give the textual file column as people using the GUI can usually work it out as the labels and descriptions help to guide and clear it up.
I'll leave it up to you what you want to do or if you want to leave as is, just wanted to point out some people speed read and dont fully take the time to understand what is being presented.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java:56]: Any reason why you dont just put the contents of this function here? You can still call this function by calling initialize() elsewhere in your code.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java:78]:
```suggestion
        if (bridge == null) {
            return;
        } 
```
I'm pretty sure that all of this is not necessary. initialize() will not get called unless: 1. you have a bridge defined. 2. the bridge is online.

The moment the bridge comes online, the framework will then call the initialize() function of all the child things. You only need to keep the compiler happy by checking if the bridge is null like I have suggested.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java:67]: I'm pretty sure all of this is not necessary. The framework gives you a lot of functionality for free, you only need to over ride it if there is a reason to and I'm not seeing the need for any of this, try removing it and testing what happens.
If your bridge goes offline, it automatically takes all the child things offline.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java:86]: Would it make sense to make this a global variable and do this in the initilise() function so it is only done once? if the parameter is ever changed the initilise() is called again by the framework.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliDeviceHandler.java:101]: What happens when the PercentType '0' command is sent, does it get handled correctly?

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java:48]: You have used this in two classes, why not put it into the DaliBindingConstants class and then use it in both?

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java:104]: What happens if the command is REFRESH or INCREASE?

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java:113]: Can you do a FIND on 'updateStatus(' and then check if you are doing a logger at the same time and change all of them please?

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java:75]: The event.log will already have this info in it because of the thingStatus changes.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java:76]:
```suggestion
        config = getConfigAs(DaliserverConfig.class);
```
This can not return null so you do not need validConfiguration() function.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java:91]: This can not occur as you have required="true" defined in the XML, so if it is empty the handler will never run. This point is moot as I believe you can remove this whole function if you use the above suggestion.

**Review by Skinah — CHANGES_REQUESTED** [review]
Ok I have gone through it all as much as I can, so once it is all addressed I will approve it and then someone else with more knowledge can do the second review. If one of my comments does not make sense feel free to say so as I am still new at Java and am self taught.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java
Comment (fwolter) [reply bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliserverBridgeHandler.java]: Thanks for your reviewing effort here and in the last time @Skinah, really appreciate it! The user could skip the check by using textual configuration.

File: bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java
Comment (rs22) [reply bundles/org.openhab.binding.dali/src/main/java/org/openhab/binding/dali/internal/handler/DaliRgbHandler.java:94]: I've now added handlers for refresh and increase/decrease commands.

**Comment by fwolter** [comment]
@rs22 Are you finished addressing or commenting @Skinah's findings?

File: bundles/org.openhab.binding.dali/README.md
Comment (rs22) [reply bundles/org.openhab.binding.dali/README.md:34]: I think I adopted this table from the skeleton generator -- and since the first column contains whitespace values (i.e. 'Port Number'), I think people will not mistake these as parameter identifiers.

**Comment by rs22** [comment]
@fwolter yes, thanks for the reminder -- I forgot to mark the requested changes as resolved. Happy Easter!

File: bundles/org.openhab.binding.dali/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.dali/src/main/resources/OH-INF/thing/thing-types.xml:56]: There are predefined system channels you could use. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#system-state-channel-types

Same for the color channel.

**Review by fwolter — CHANGES_REQUESTED** [review]
Thanks for your contribution! I reviewed your code and here is my feedback.

Please add yourself to the `CODEOWNERS` file and the binding to `bom\openhab-addons\pom.xml`.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

File: bundles/org.openhab.binding.dali/README.md
Comment (rs22) [reply bundles/org.openhab.binding.dali/README.md]: I removed the custom channels from the binding entirely.

**Comment by rs22** [comment]
Thanks for your helpful reviews! I hope that I could address everything with my latest updates.

**Review by fwolter — APPROVED** [review]
LGTM

Now, you could add a logo: https://deploy-preview-1540--openhab-docs-preview.netlify.app/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

**Pull request merged**
